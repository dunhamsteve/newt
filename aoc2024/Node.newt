../aoc2023/Node.newt