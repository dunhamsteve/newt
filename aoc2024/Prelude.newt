../aoc2023/Prelude.newt