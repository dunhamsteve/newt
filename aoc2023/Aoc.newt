module Aoc

import Prelude

nums : String → List Int
nums s = map stringToInt $ filter (_/=_ "") $ split (trim s) " "

isDigit : Char -> Bool
isDigit '0' = True
isDigit '1' = True
isDigit '2' = True
isDigit '3' = True
isDigit '4' = True
isDigit '5' = True
isDigit '6' = True
isDigit '7' = True
isDigit '8' = True
isDigit '9' = True
isDigit _ = False

indexOf? : ∀ a. {{Eq a}} → a → List a → Maybe Nat
indexOf? {a} z xs = go Z z xs
  where
    go : Nat → a → List a → Maybe Nat
    go ix z Nil = Nothing
    go ix z (x :: xs) =
      if z == x then Just ix else go (S ix) z xs

-- if_then_else shorthand
-- Lean version uses a decidable instead of Bool
ite : ∀ a. Bool → a → a → a
ite c t e = if c then t else e

-- probably not super efficient, but it works
qsort : ∀ a. (a → a → Bool) → List a → List a
qsort lt Nil = Nil
qsort lt (x :: xs) = qsort lt (filter (λ y => not $ lt x y) xs) ++ x :: qsort lt (filter (lt x) xs)
