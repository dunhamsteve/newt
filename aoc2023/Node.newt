module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv)`
pfunc readFile uses (fs MkIORes Unit) : (fn : String) -> IO String := `(fn) => (w) => MkIORes(Unit, fs.readFileSync(fn, 'utf8'), w)`
