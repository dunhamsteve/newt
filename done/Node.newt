module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(undefined, undefined, content)
  } catch (e) {
    let err = ""+e
    result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(undefined, undefined, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

