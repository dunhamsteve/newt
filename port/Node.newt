module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(null, null, content)
  } catch (e) {
    let err = ""+e
    result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(null, null, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

