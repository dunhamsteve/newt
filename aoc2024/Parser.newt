module Parser

import Prelude
import Aoc

Parser : U → U
Parser a = List Char → Either String (a × List Char)

instance Monad Parser where
  pure a = \ cs => Right (a, cs)
  bind ma mab = \ cs => ma cs >>= uncurry mab

instance Alternative Parser where
  pa <|> pb = \ cs => case pa cs of
    Left msg => pb cs
    res => res

instance Functor Parser where
  map f pa = \ cs => case pa cs of
    Left msg => Left msg
    Right (a, cs) => Right (f a, cs)

instance Applicative Parser where
  return a = pure a
  pa <*> pb = pa >>= (\ f => map f pb)


fail : ∀ a. String -> Parser a
fail msg = \ cs => Left msg

-- TODO, case builder isn't expanding Parser Unit to count lambdas
eof : Parser Unit
eof = \ cs => case cs of
  Nil => Right (MkUnit, Nil)
  _ =>  Left "expected eof"

satisfy : (Char → Bool) → Parser Char
satisfy pred = λ cs => case cs of
  Nil => Left "unexpected EOF"
  (c :: cs) => if pred c then Right (c, cs) else Left ("did not expect " ++ show c)

match : Char → Parser Char
match d = satisfy (_==_ d)

any : Parser Char
any = satisfy (λ _ => True)

some many : ∀ a. Parser a → Parser (List a)
many p = some p <|> pure Nil
some p = do
  v <- p
  vs <- many p
  pure (v :: vs)

string : String → Parser Unit
string str = go (unpack str)
  where
    go : List Char → Parser Unit
    go Nil = pure MkUnit
    go (c :: cs) = match c >> go cs

number : Parser Int
number = stringToInt ∘ pack <$> some (satisfy isDigit)
  -- do
  -- digs <- some (satisfy isDigit)
  -- pure $ stringToInt $ pack digs

ws : Parser Unit
ws = many (match ' ') >> pure MkUnit
