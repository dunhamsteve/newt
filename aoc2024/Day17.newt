module Day17

import Prelude
import Node
import Aoc
import Parser

ptype BigInt

pfunc itobi : Int → BigInt := `(x) => BigInt(x)`
pfunc bitoi : BigInt → Int := `(x) => Number(x)`
pfunc addbi : BigInt → BigInt → BigInt := `(a,b) => a + b`
pfunc subbi : BigInt → BigInt → BigInt := `(a,b) => a - b`
pfunc mulbi : BigInt → BigInt → BigInt := `(a,b) => a * b`
pfunc divbi : BigInt → BigInt → BigInt := `(a,b) => a / b`

instance Mul BigInt where a * b = mulbi a b
instance Div BigInt where a / b = divbi a b
instance Add BigInt where a + b = addbi a b
instance Sub BigInt where a - b = subbi a b
instance Cast Int BigInt where cast x = itobi x
instance Eq BigInt where a == b = jsEq a b
instance Show BigInt where show = jsShow
instance Ord BigInt where compare a b = jsCompare a b

data Machine : U where
  M : BigInt → BigInt → BigInt → List Int → Int → SnocList Int → Machine

preg : Parser BigInt
preg = do
  token "Register"
  any
  token ":"
  n <- number
  match '\n'
  pure $ itobi n

pprog : Parser (List Int)
pprog = do
  token "Program:"
  n <- number
  ns <- many $ match ',' >> number
  many $ match '\n'
  pure $ n :: ns

pmachine : Parser Machine
pmachine = do
  a <- preg
  b <- preg
  c <- preg
  many $ match '\n'
  prog <- pprog
  pure $ M a b c prog 0 Lin

infixl 7 _**_ _|_ _^_ _%_ _>>>_
pfunc _**_ : Int → Int → Int := `(x,y) => x ** y`
pfunc _>>>_ : BigInt → BigInt → BigInt := `(x,y) => x >> y`
pfunc _%_ : BigInt → BigInt → BigInt := `(x,y) => x % y`
pfunc bxor : BigInt → BigInt → BigInt := `(x,y) => x ^ y`

step : Machine -> Machine
step mach@(M a b c mem ip out) =
  let ip' = cast ip in
  let (Just ins) = getAt ip' mem | _ => mach in
  let (Just op) = getAt (S ip') mem | _ => mach in
  case ins of
    0 => let a = a >>> combo op in step (M a b c mem (ip + 2) out)
    1 => let b = bxor b (itobi op) in step (M a b c mem (ip + 2) out)
    2 => let b = combo op % itobi 8 in step (M a b c mem (ip + 2) out)
    3 => if a == itobi 0 then step (M a b c mem (ip + 2) out) else step (M a b c mem op out)
    4 => let b = bxor b c in step (M a b c mem (ip + 2) out)
    5 => let o = combo op % itobi 8 in step (M a b c mem (ip + 2) (out :< bitoi o))
    6 => let b = a >>> combo op in step (M a b c mem (ip + 2) out)
    7 => let c = a >>> combo op in step (M a b c mem (ip + 2) out)
    _ => mach
  where
    combo : Int → BigInt
    combo 4 = a
    combo 5 = b
    combo 6 = c
    combo n = itobi n

Cand : U
Cand = BigInt ×  SnocList Int ×  List Int

part2 : Machine -> Maybe BigInt
part2 mach@(M a b c mem ip out) = go ((itobi 0,(Lin <>< mem), Nil) :: Nil)
  where
    cands : List BigInt
    cands = map itobi (0 :: 1 :: 2 :: 3 :: 4 :: 5 :: 6 :: 7 :: Nil)

    matches : List Int → BigInt → Bool
    matches want a =
      let (M _ b c mem ip out) = step (M a b c mem ip Lin) in
      let out = out <>> Nil in
      out == want

    chain : (Int → Maybe Int) → List Int → Maybe Int
    chain f Nil = Nothing
    chain f (x :: xs) = case f x of
      Just a => Just a
      Nothing => chain f xs

    go : List Cand → Maybe BigInt
    go Nil = Nothing
    go ((a, Lin, acc) :: _) = Just a
    go ((a, xs :< x, acc) :: todo) =
      let next = filter (matches (x :: acc)) $ map (_+_ (a * itobi 8)) cands in
      go (map (\a => (a, xs, x :: acc)) next ++ todo)

render : SnocList Int → String
render out = go out Nil
  where
    go : SnocList Int → List Char → String
    go Lin chars = pack $ tail chars
    go (xs :< x) acc = go xs (',' :: chr (x + 48) :: acc)

run : String -> IO Unit
run fn = do
  putStrLn fn
  text <- readFile fn

  let (Right (mach, Nil)) = pmachine (unpack text)
    | (Right (_, cs)) => putStrLn $ "extra: " ++ pack cs
    | _ => putStrLn "parse error"

  let (M a b c mem ip out) = (step mach)
  putStrLn $ "part1 " ++ render out
  let (Just p2) = part2 mach | Nothing => putStrLn "fail"
  putStrLn  $ "part2 " ++ show p2
  putStrLn ""

main : IO Unit
main = do
  run "aoc2024/day17/eg.txt"
  run "aoc2024/day17/eg2.txt"
  run "aoc2024/day17/input.txt"
