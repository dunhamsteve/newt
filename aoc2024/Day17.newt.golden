aoc2024/day17/eg.txt
part1 4,6,3,5,6,3,5,2,1,0
fail
aoc2024/day17/eg2.txt
part1 5,7,3,0
part2 117440

aoc2024/day17/input.txt
part1 5,0,3,5,7,6,1,5,4
part2 164516454365621

